pkg/api/converterfromv20180930preview.go

```diff
@@ -6,92 +6,314 @@ import (
+		}
+		// If the requested profile does not exist, add it
+		// in cs as is, otherwise merge it in the existing
+		// profile.
+		if index == -1 {
+			cs.Properties.AgentPoolProfiles = append(cs.Properties.AgentPoolProfiles, convertAgentPoolProfile(in))
+		} else {
+			if in.Count != nil {
+				cs.Properties.AgentPoolProfiles[index].Count = *in.Count
+			}
+			if in.VMSize != nil {
+				cs.Properties.AgentPoolProfiles[index].VMSize = VMSize(*in.VMSize)
+			}
+			if in.SubnetCIDR != nil {
+				cs.Properties.AgentPoolProfiles[index].SubnetCIDR = *in.SubnetCIDR
+			}
+			if in.OSType != nil {
+				cs.Properties.AgentPoolProfiles[index].OSType = OSType(*in.OSType)
+			}
+			if in.Role != nil {
+				cs.Properties.AgentPoolProfiles[index].Role = AgentPoolProfileRole(*in.Role)
+			}
+		}
+	}
+}
+
+func agentPoolProfileIndex(name string, profiles []AgentPoolProfile) int {
+	for i, profile := range profiles {
+		if profile.Name == name {
+			return i
+		}
+	}
+	return -1
+}
+
+func convertMasterToAgentPoolProfile(in v20180930preview.MasterPoolProfile) (out AgentPoolProfile) {
+	out.Name = string(AgentPoolProfileRoleMaster)
+	out.Role = AgentPoolProfileRoleMaster
+	out.OSType = OSTypeLinux
+	if in.Count != nil {
+		out.Count = *in.Count
+	}
+	if in.VMSize != nil {
+		out.VMSize = VMSize(*in.VMSize)
+	}
+	if in.SubnetCIDR != nil {
+		out.SubnetCIDR = *in.SubnetCIDR
+	}
+	return
+}
+
+func convertAgentPoolProfile(in v20180930preview.AgentPoolProfile) (out AgentPoolProfile) {
+	if in.Name != nil {
+		out.Name = *in.Name
+	}
+	if in.Count != nil {
+		out.Count = *in.Count
+	}
+	if in.VMSize != nil {
+		out.VMSize = VMSize(*in.VMSize)
+	}
+	if in.SubnetCIDR != nil {
+		out.SubnetCIDR = *in.SubnetCIDR
+	}
+	if in.OSType != nil {
+		out.OSType = OSType(*in.OSType)
+	}
+	if in.Role != nil {
+		out.Role = AgentPoolProfileRole(*in.Role)
+	}
+	return
+}
+
+func mergeAuthProfile(oc *v20180930preview.OpenShiftManagedCluster, cs *OpenShiftManagedCluster) {
+	if oc.Properties.AuthProfile == nil {
+		return
+	}
+
+	if cs.Properties.AuthProfile.IdentityProviders == nil {
+		cs.Properties.AuthProfile.IdentityProviders = make([]IdentityProvider, 0, len(oc.Properties.AuthProfile.IdentityProviders))
+	}
+
+	for _, ip := range oc.Properties.AuthProfile.IdentityProviders {
+		index := -1
+		if ip.Name != nil {
+			index = identityProviderIndex(*ip.Name, cs.Properties.AuthProfile.IdentityProviders)
+		}
+		// If the requested provider does not exist, add it
+		// in cs as is, otherwise merge it in the existing
+		// provider.
+		if index == -1 {
+			cs.Properties.AuthProfile.IdentityProviders = append(cs.Properties.AuthProfile.IdentityProviders, convertIdentityProvider(ip))
+		} else {
+			provider := cs.Properties.AuthProfile.IdentityProviders[index].Provider
+			switch out := provider.(type) {
+			case (*AADIdentityProvider):
+				in := ip.Provider.(*v20180930preview.AADIdentityProvider)
+				if in.Kind != nil {
+					out.Kind = *in.Kind
```

Comment:
I think this should not be updatable/changeable.

Reply:
Well, for starters, it needs to be set on creation. Whether it's valid or not to update should be handled in validation.

Reply:
true. I think we will need to shape these edge cases one by one.